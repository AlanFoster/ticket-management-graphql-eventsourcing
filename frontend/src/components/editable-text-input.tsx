import * as React from "react";
import { useState } from "react";
import { InputBase, makeStyles } from "@material-ui/core";
import { fade } from "@material-ui/core/styles";

const useStyles = makeStyles(theme => ({
    text: {
        ...theme.typography.h5,
        border: 0,
        padding: '0.5rem',
        borderRadius: '5px',
        width: '100%',
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        textOverflow: 'ellipse',
        height: 'inherit',
        '&:hover': {
            transition: theme.transitions.create(['background-color', 'box-shadow'], { duration: 300 }),
            backgroundColor: '#f1f1f0',
        },
        '&:focus': {
            backgroundColor: '#fcfcfb',
            boxShadow: `${fade(theme.palette.primary.main, 0.25)} 0 0 0 2px`,
        },
    },

}));

export const EditableTextInput = function (props) {
    const classes = useStyles();
    const [isEditing, setEditing] = useState(false);
    const [value, setValue] = useState(props.value);

    const saveField = function () {
        const hasValueChanged = value !== props.value;
        if (hasValueChanged) {
            props.onChange(value);
        } else {
            setEditing(false);
        }
    };

    const onKeyUp = function (e) {
        if (e.key === 'Enter') {
            saveField();
        }
    };

    return (
        <React.Fragment>
            {!isEditing &&
                <div
                    className={classes.text}
                    onClick={() => setEditing(true)}
                >
                    {props.value}
                </div>
            }
            {isEditing &&
            <InputBase
                value={value}
                onChange={e => setValue(e.target.value)}
                onKeyPress={onKeyUp}
                classes={{
                    input: classes.text,
                }}
                inputProps={{
                    onBlur: saveField,
                }}
                required
                fullWidth={true}
                autoFocus
            />
            }
        </React.Fragment>
    )
};
