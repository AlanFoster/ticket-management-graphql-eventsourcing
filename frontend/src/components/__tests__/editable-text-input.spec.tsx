import * as React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react";
import { EditableTextInput } from "../editable-text-input";

describe("EditableText", function() {
    afterEach(cleanup);

    it("shows the given initial value", function() {
        const onChange = jest.fn();
        const { container } = render(
            <EditableTextInput value="Initial value" onChange={onChange} />
        );

        expect(onChange).toHaveBeenCalledTimes(0);
        expect(container).toMatchSnapshot();
    });

    it("can be edited", function() {
        const onChange = jest.fn();
        const { getByText, getByDisplayValue, container } = render(
            <EditableTextInput value="Initial value" onChange={onChange} />
        );

        fireEvent.click(getByText(/initial value/i));
        fireEvent.change(getByDisplayValue(/initial value/i), {
            target: { value: "new value" }
        });
        fireEvent.blur(getByDisplayValue(/new value/i));

        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledWith("new value");
        expect(container).toMatchSnapshot();
    });
});
