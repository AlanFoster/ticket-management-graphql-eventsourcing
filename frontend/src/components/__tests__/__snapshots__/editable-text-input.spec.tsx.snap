// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableText can be edited 1`] = `
<div>
  <div
    class="MuiInputBase-root MuiInputBase-fullWidth Mui-focused"
    required=""
  >
    <input
      class="MuiInputBase-input makeStyles-text-2"
      required=""
      type="text"
      value="new value"
    />
  </div>
</div>
`;

exports[`EditableText shows the given initial value 1`] = `
<div>
  <div
    class="makeStyles-text-1"
  >
    Initial value
  </div>
</div>
`;
